cloud/blockstore/libs/storage/disk_agent/disk_agent_actor_direct_copy.cpp

```diff
@@ -0,0 +1,218 @@
+    return true;
+}
+
+void TDirectCopyActor::Done(const NActors::TActorContext& ctx)
+{
+    NCloud::Reply(
+        ctx,
+        *RequestInfo,
+        std::make_unique<TEvDiskAgent::TEvDirectCopyBlocksResponse>(
+            MakeError(S_OK)));
+
+    Die(ctx);
+}
+
+void TDirectCopyActor::HandleReadBlocksResponse(
+    const TEvDiskAgent::TEvReadDeviceBlocksResponse::TPtr& ev,
+    const TActorContext& ctx)
+{
+    auto* msg = ev->Get();
+
+    if (HandleError(ctx, msg->GetError())) {
+        return;
+    }
+
+    auto writeRequest =
+        std::make_unique<TEvDiskAgent::TEvWriteDeviceBlocksRequest>();
+
+    auto& rec = writeRequest->Record;
+    auto* headers = rec.MutableHeaders();
+    headers->SetIsBackgroundRequest(Request.GetHeaders().GetIsBackgroundRequest());
+    headers->SetClientId(TString(Request.GetTargetClientId()));
+
+    rec.MutableBlocks()->Swap(msg->Record.MutableBlocks());
+    rec.SetDeviceUUID(Request.GetTargetDeviceUUID());
+    rec.SetStartIndex(Request.GetTargetStartIndex());
+    rec.SetBlockSize(Request.GetBlockSize());
+
+    auto targetActorId =
+        ctx.SelfID.NodeId() == Request.GetTargetNodeId()
+            ? MakeDiskAgentServiceId()
+            : MakeDiskAgentServiceId(Request.GetTargetNodeId());
+    auto event = std::make_unique<IEventHandle>(
+        targetActorId,
+        ctx.SelfID,
+        writeRequest.release(),
+        IEventHandle::FlagForwardOnNondelivery,
+        0,            // Cookie
+        &ctx.SelfID   // forwardOnNondelivery
+    );
+
+    ctx.Send(event.release());
+}
+
+void TDirectCopyActor::HandleWriteBlocksResponse(
+    const TEvDiskAgent::TEvWriteDeviceBlocksResponse::TPtr& ev,
+    const TActorContext& ctx)
+{
+    auto* msg = ev->Get();
+
+    if (HandleError(ctx, msg->GetError())) {
+        return;
+    }
+
+    Done(ctx);
+}
+
+void TDirectCopyActor::HandleWriteBlocksUndelivery(
+    const TEvDiskAgent::TEvWriteDeviceBlocksRequest::TPtr& ev,
+    const TActorContext& ctx)
+{
+    Y_UNUSED(ev);
+
+    HandleError(ctx, MakeError(E_REJECTED, "WriteBlocks request undelivered"));
+}
+
+STFUNC(TDirectCopyActor::StateWork)
+{
+    switch (ev->GetTypeRewrite()) {
+        HFunc(
+            TEvDiskAgent::TEvReadDeviceBlocksResponse,
+            HandleReadBlocksResponse);
+
+        HFunc(
+            TEvDiskAgent::TEvWriteDeviceBlocksResponse,
+            HandleWriteBlocksResponse);
+        HFunc(
+            TEvDiskAgent::TEvWriteDeviceBlocksRequest,
+            HandleWriteBlocksUndelivery);
+
+        default:
+            HandleUnexpectedEvent(ev, TBlockStoreComponents::DISK_AGENT_WORKER);
+            break;
+    }
+}
+
+}   // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+
+void TDiskAgentActor::HandleDirectCopyBlocks(
+    const TEvDiskAgent::TEvDirectCopyBlocksRequest::TPtr& ev,
+    const TActorContext& ctx)
+{
+    const auto* msg = ev->Get();
+    const auto& record = msg->Record;
+
+    LOG_INFO(
```

Comment:
Will INFO logging here be too verbose especially if we direct copy small chunks of data ?
Additionally let's add the data range and offsets